spot-client/src/common/logger/post-to-endpoint.js

```diff
@@ -54,19 +96,76 @@ export default class PostToEndpoint {
             }
         }, []);
 
-        return fetch(
-            this._endpointUrl,
-            {
-                body: JSON.stringify({
-                    events: jsonEvents,
-                    'device_id': this._deviceId
-                }),
-                headers: {
-                    'Content-Type': 'application/json'
-                },
-                method: 'POST',
-                mode: 'cors'
-            }
-        );
+        this._requestsQueue.push(
+            new PostLogsRequest(/* attempt */ 1, jsonEvents));
+
+        // Make sure that there's only 1 request on the flight by starting
+        // the reaction chain only if there's 1 request in the queue (the one
+        // added above).
+        if (this._requestsQueue.length === 1) {
+            this._sendLogs();
+        }
+    }
+
+    /**
+     * Tries to send first log request from the requests queue. In case of
+     * success it will keep taking requests of the queue until it's empty.
+     * On failure it will retry the request with a delay until
+     * the {@link PostToEndpoint#MAX_ATTEMPTS} is reached.
+     *
+     * @returns {void}
+     * @private
+     */
+    _sendLogs() {
+        const request = this._requestsQueue.shift();
```

Comment:
Thought of a race here: if you pull out the logs now the array could be empty while the fetch is in flight. If logs get queued then another fetch could get fired.

Reply:
Yup, good catch. So maybe the request should not be removed from the queue until it gets processed ?

Reply:
Yeah. And that way you also don't have to re-add the payload if it fails.